src/main/java/org/topbraid/shacl/tools/BlankNodeFinder.java

```diff
@@ -0,0 +1,63 @@
+package org.topbraid.shacl.tools;
+
+import org.apache.jena.rdf.model.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.topbraid.shacl.vocabulary.SH;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Stream;
+
+public class BlankNodeFinder {
+    private static final Logger logger = LoggerFactory.getLogger(BlankNodeFinder.class);
+    /**
+     * Finds all blank nodes in the report that references the shapes graph
+     * @param report The validation report
+     * @param shapesGraph The shapes graph
+     * @return A model containing all blank nodes
+     */
+    public static Model findBlankNodes(Model report, Model shapesGraph) {
+        // Find all blank nodes that references the shapes graph
+
+        Model blankNodes = ModelFactory.createDefaultModel();
+
+        Stream<Statement> statements = report.listStatements(null, SH.result, (RDFNode) null).toList().stream();
```

Comment:
Just a minor comment here: this first creates a List and then a Stream. This may not scale well for large reports. Better would be to use an ExtendedIterator<Statement> instead, which also has a map function

Reply:
Changed it all list creation to use only iterators